WIP: Changes BytesMut not to rely on Vec + Arc

Matthias247 opened the pull request:
This changes the internal BytesMut representation to allocate memory on
its own, and not to use Arc + Vec under the hood.

Doing this allows to have a single representation which is shareable (by
being able to modify an embedded refcount) with zero-cost right from
the start. Not representation changes have to be performed over the
lifetime of `BytesMut`. Therefore `BytesMut::freeze()` is now also free.

The change also opens the door for adding support for buffer pools or
custom allocators, since an allocator is directly called. Adding this
support would mostly mean having to pass an allocator to BytesMut
which is used called instead of a fixed function. This would however be
a different change.

The change is WIP, and misses the following parts:
- The buffer growth strategy is not exactly the same as in the old
  version. It needs to be reviewed what the best strategy is.
- Some situations where an empty buffer (which has no backing
  storage) is manipulated might have bugs. This needs another
  round of review, and some additional tests.

Matthias247 commented on src/bytes_mut.rs:
```diff
@@ -602,65 +533,60 @@ impl BytesMut {
+            // There are some situations, using `reserve_exact` that the
+            // buffer capacity could be below `original_capacity`, so do a
+            // check.
+            let double = self.cap.checked_shl(1).unwrap_or(new_cap);
```
This is one part that leads to the failing tests.
The old version doubled the original capacity of the storage. However if that was split, the capacity of the split part could be much lower.

I'm not really sure whether going back to the original capacity or even doubling it makes sense => If someone split off a 100 byte chunk of a 1MB `BytesMut` and they want to write a few more bytes into it, it shouldn't grow towards 1 or even 2 MB - right?

I'm somehwat leaning towards ignoring the original capacity at all, just making decisions based on capacity of the current part. But maybe I'm just missing why it was added that way? Maybe to make sure that a receive buffer somewhere which uses `BytesMut` and from deserialized data is split off always stays at the same size? If people would want that, they could still `.reserve()` that size themselves after splitting.

camshaft replied:
I agree that it seems like surprising behavior. I would vote to ignore the original capacity.

Darksonn commented on src/bytes_mut.rs:
```diff
@@ -139,7 +103,25 @@ impl BytesMut {
+            if capacity == 0 {
+                BytesMut {
+                    data: ptr::null_mut(),
+                    ptr: NonNull::new_unchecked(&GUARD as *const u8 as *mut u8),
```
Why not just `NonNull::dangling`?

Matthias247 replied:
I just wasn't aware of it. Updated to use that now instead

Darksonn commented on src/bytes_mut.rs:
```diff
@@ -63,50 +63,14 @@ pub struct BytesMut {
-const PTR_WIDTH: usize = 64;
-#[cfg(target_pointer_width = "32")]
-const PTR_WIDTH: usize = 32;
+/// An element in static memory which we use as guard pointer for
+/// empty byte arrays
+const GUARD: u8 = 0;
```
This sounds like it should be a `static`?

carllerche commented:
cc/ @seanmonstar who did most of this work most recently.

Matthias247 commented:
I'm now rather happy with this, so I removed the `WIP`.

I added some more documentation around the memory allocation strategy - based on my understanding of previous versions - and questions on what we should do there.
I'm somewhat inclined to drop the behavior of taking capacity that an old backing storage had into account for reservations - it just seems like a footgun leading to high memory usage in case people split `BytesMut` and work on them later in individually.

Matthias247 commented:
Not sure what is going on with Loom. I think it doesn't like my use of `alloc::alloc::alloc` (direct use of the global allocator)- but don't know how to fix it.

Matthias247 commented:
`cargo bench` comparison:

```
name                      old ns/iter      new ns/iter      diff ns/iter   diff %  speedup 
 alloc_big                 52               53                          1    1.92%   x 0.98
 alloc_mid                 53               51                         -2   -3.77%   x 1.04
 alloc_small               53,018           51,740                 -1,278   -2.41%   x 1.02
 alloc_write_split_to_mid  104              67                        -37  -35.58%   x 1.55
 bytes_mut_extend          568 (225 MB/s)   538 (237 MB/s)            -30   -5.28%   x 1.06 
 clone_frozen              9,711            9,785                      74    0.76%   x 0.99
 deref_shared              242              241                        -1   -0.41%   x 1.00
 deref_two                 237              235                        -2   -0.84%   x 1.01
 deref_unique              239              240                         1    0.42%   x 1.00
 deref_unique_unroll       257              258                         1    0.39%   x 1.00
 drain_write_drain         298              228                       -70  -23.49%   x 1.31 
 fmt_write                 14 (2642 MB/s)   14 (2642 MB/s)              0    0.00%   x 1.00
 put_slice_bytes_mut       15 (8533 MB/s)   14 (9142 MB/s)             -1   -6.67%   x 1.07
 put_slice_vec             16 (8000 MB/s)   15 (8533 MB/s)             -1   -6.25%   x 1.07
 put_slice_vec_extend      10 (12800 MB/s)  10 (12800 MB/s)             0    0.00%   x 1.00
 put_u8_bytes_mut          508 (251 MB/s)   454 (281 MB/s)            -54  -10.63%   x 1.12
 put_u8_vec                538 (237 MB/s)   513 (249 MB/s)            -25   -4.65%   x 1.05
 put_u8_vec_push           128 (1000 MB/s)  127 (1007 MB/s)            -1   -0.78%   x 1.01
```

camshaft commented on src/bytes_mut.rs:
```diff
@@ -823,12 +761,11 @@ impl BytesMut {
+        // Note that the following comparison will also yield true for 2 empty
+        // BytesMut. This will however be fine - it will simply combine them
+        // into a single empty BytesMut.
         let ptr = unsafe { self.ptr.as_ptr().offset(self.len as isize) };
```
From https://doc.rust-lang.org/std/primitive.pointer.html#safety-2:

> If any of the following conditions are violated, the result is Undefined Behavior:
> ...
> * The computed offset, in bytes, cannot overflow an isize.
> ...
> The compiler and standard library generally tries to ensure allocations never reach a size where an offset is a concern. For instance, Vec and Box ensure they never allocate more than isize::MAX bytes, so vec.as_ptr().add(vec.len()) is always safe.
>
> Most platforms fundamentally can't even construct such an allocation. For instance, no known 64-bit platform can ever serve a request for 2^63 bytes due to page-table limitations or splitting the address space. However, some 32-bit and 16-bit platforms may successfully serve a request for more than isize::MAX bytes with things like Physical Address Extension. As such, memory acquired directly from allocators or memory mapped files may be too large to handle with this function.

Do we put a cap on `len` anywhere to make sure this holds?

Matthias247 replied:
I don't think we currently do. But we can limit the maximum capacity to `isize::MAX` in `reserve_inner`

camshaft commented on src/bytes_mut.rs:
```diff
@@ -1246,24 +1092,70 @@ impl Shared {
+
+    unsafe fn increment_refcount(self_ptr: *mut Self) {
+        debug_assert!(!self_ptr.is_null());
+        let old_size = (*self_ptr).refcount.fetch_add(1, Ordering::Relaxed);
```
Can we make a note about why it's ok to be `Relaxed` here?

From https://doc.rust-lang.org/nomicon/atomics.html#relaxed

> Relaxed operations are appropriate for things that you definitely want to happen, but don't particularly otherwise care about. For instance, incrementing a counter can be safely done by multiple threads using a relaxed fetch_add if you're not using the counter to synchronize any other accesses.

Also https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering
> Typical use for relaxed memory ordering is incrementing counters, such as the reference counters of std::shared_ptr, since this only requires atomicity, but not ordering or synchronization (note that decrementing the shared_ptr counters requires acquire-release synchronization with the destructor)

Matthias247 replied:
Is "copied from the other refcounted object implementations" a good enough comment? 😀

My understanding is that `Relaxed` is good here because we don't mutate or watch any data apart from the refcount when this is called.

camshaft replied:
I would be fine if we linked and inlined either/both of those links

camshaft commented on src/bytes_mut.rs:
```diff
@@ -1246,24 +1092,70 @@ impl Shared {
     }
 
-    1 << (repr + (MIN_ORIGINAL_CAPACITY_WIDTH - 1))
+    unsafe fn data_ptr_mut(&mut self) -> NonNull<u8> {
+        let mut end_addr = self as *const Shared as usize;
+        end_addr += mem::size_of::<Shared>();
+        NonNull::new_unchecked(end_addr as *mut u8)
+    }
```
nit: i feel like it's a bit cleaner to use `offset`:

```suggestion
    unsafe fn data_ptr_mut(&mut self) -> NonNull<u8> {
        let shared_addr = self as *mut Self;
        let data_addr = shared_addr.offset(1) as *mut u8;
        NonNull::new_unchecked(data_addr)
    }
```

Matthias247 replied:
Maybe. It should lead to the same result, but I'm not sure whether I should trust Rust more or less about casting the pointer at the end of `Shared` to a `u8` than at the beginning :)

camshaft replied:
They are the same result, for sure. I guess I was more concerned about the number of conversions that occur in the previous version, especially the `&mut` -> `*const` -> `*mut`:

`&mut Shared` -> `*const Shared` -> `usize` `*mut u8`

vs

`&mut Shared` -> `*mut Shared` -> `*mut u8`

Darksonn replied:
Don't cast a `&mut T` to a `*const T`. It involves first coercing it to an `&T`, which means that it would be unsound to perform any writes through the resulting pointer. Try running [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=da1be247e82bc90ae919153b9c9e05ee) using miri from the tools.

camshaft commented on src/bytes_mut.rs:
```diff
@@ -292,6 +275,12 @@ impl BytesMut {
         );
+
+        if self.data.is_null() {
+            debug_assert_eq!(at, 0, "Empty Bytes is only splittable at the front");
```
Is this possible to ever violate? since when `data.is_null()` then `capacity = 0` and the above assertion will check that?

Matthias247 replied:
It shouldn't be possible to violate it. It's more just a guard and indicator in case of something was missed somewhere, or future changes are added. That's also why it is a `debug_assert` and not an `assert`

camshaft replied:
Makes sense

Matthias247 commented on src/bytes_mut.rs:
```diff
@@ -139,7 +105,25 @@ impl BytesMut {
     #[inline]
     pub fn with_capacity(capacity: usize) -> BytesMut {
-        BytesMut::from_vec(Vec::with_capacity(capacity))
+        unsafe {
```
Right . I think we could actually make `Shared::allocate_for_size` safe. There's nothing bad about calling it. The unsafe parts are rather about interacting with it later

Matthias247 commented on src/bytes_mut.rs:
```diff
@@ -763,6 +717,12 @@ impl BytesMut {
 
     #[inline]
     fn as_slice_mut(&mut self) -> &mut [u8] {
+        // TODO:
+        // The pointer is always valid, but might not backed by real storage
+        // in case of an empty BytesMut.
+        // This shouldn't matter due to the array length being 0 and the pointer
+        // never being accessed. However it is still not fully clear if this is
+        // valid within Rusts's UB rules.
```
Yeah, that's my understanding too. I just wanted someone to double check.
Then we can just remove that comment and TODO

Matthias247 commented on src/bytes_mut.rs:
```diff
@@ -1197,44 +1067,20 @@ impl<'a> FromIterator<&'a u8> for BytesMut {
-unsafe fn increment_shared(ptr: *mut Shared) {
-    let old_size = (*ptr).ref_count.fetch_add(1, Ordering::Relaxed);
+impl Shared {
+    unsafe fn allocate_for_size(size: usize) -> Result<*mut Shared, ()> {
```
As mentioned elsewhere, it shouldn't be unsafe to call. It however uses lots of `unsafe` internal. We can certainly move that around.

carllerche commented:
It looks like you are inlining the capacity with the data? Given that buffers are often allocated to be page length, would this not result in odd allocation patterns? The size would often be `size_of::<Shared>() + PAGE`.

carllerche commented:
Looking through this, I believe this should be able to land w/o a breaking change. I don't think we need to block 0.3 on this.

Matthias247 commented:
> It looks like you are inlining the capacity with the data? Given that buffers are often allocated to be page length, would this not result in odd allocation patterns? The size would often be `size_of::<Shared>() + PAGE`.

Are you aware about any applications which do that? I don't think you can currently rely on that, since your system memory allocator might also not do the same - e.g. it might also need to store it's metadata inside the allocation header and therefore take size away from it. But it's true that buffer sizes which are multiples of 4k are often programmers favorites.

What we could do is maybe rounding up allocations to page size, so that the capacity isn't lost? But that might require adding a `with_exact_capacity` method for use-cases which don't want that.

Matthias247 commented:
regarding 

> It looks like you are inlining the capacity with the data? Given that buffers are often allocated to be page length, would this not result in odd allocation patterns? The size would often be size_of::<Shared>() + PAGE.

I confirmed that e.g. mimalloc will optimize fixed sized chunks, and `Shared` would might mean getting pushed to the next chunk size.

Based on this I see 2 ways forward for this:
1. Don't care about it in the normal case, and add more optimized APIs for people that care about the differences.E.g. `BytesMut::with_exact_capacity(size)` would return exactly what the user requires (even it if it is not optimal), `BytesMut::with_optimized_capacity(size)` could either round up (to next multiple of 2) or down (subtract `- size_of::<Shared>`).
2. Add another pointer to `Bytes` itself which tracks the shared state separately from the data. The benefit of this is that this could also eliminate the special casing for `Vec` in `Bytes`, since the field can be used to store the capacity. The downsides are however:
  - 2 allocations for `BytesMut` instead of.
  - 2 times the pointer chasing on resizes/frees/etc.

I guess for the use-case of "have a small serialization buffer and continuously add stuff" approach 2) will end up slower than 1) (and this CR). However it's more deterministic for managing large chunks of data.

carllerche commented:
Do we know why we can't entirely hide the the `Arc` / `Vec` implementation behind the vtable? IIRC it is due to wanting to avoid an allocation when converting `Vec` and needing to track the pointer to the head of the memory as well as the original capacity? Is there another reason that I am forgetting?

qm3ster commented:
Would this also allow for free `Bytes::try_mut()`? #368

saethlin commented on src/bytes_mut.rs:
```diff
@@ -1246,76 +1094,107 @@ impl Shared {
+    /// returned from `Shared::allocate_for_size` which which have not been
+    /// freed using `release_refcount`.
+    unsafe fn data_ptr_mut(&mut self) -> NonNull<u8> {
+        let shared_addr = self as *mut Self;
+        let data_addr = shared_addr.offset(1) as *mut u8;
```
FYI, under Stacked Borrows this pointer only has provenance over `Self`, so using it to access the data section would be UB. Under PNVI (the proposed C aliasing model) I believe this is allowed. It's not entirely clear what the finalized rules for Rust will be, but you should probably do one of:

* Fix this under SB, probably by adding a `[u8]` member to `Shared`
* Fix this when it becomes UB for sure
* Campaign for this to not become UB, perhaps on https://github.com/rust-lang/unsafe-code-guidelines/issues/256
